File: app/Models/Product.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Facades\Storage;

/**
 * @property int $id
 * @property int|null $category_id
 * @property int|null $product_category_id
 * @property string $name
 * @property string $slug
 * @property string|null $sku
 * @property string|null $description
 * @property string|null $short_description
 * @property string|null $content
 * @property string|null $ingredients
 * @property string|null $age_group
 * @property string|null $type
 * @property string|null $price
 * @property string|null $sale_price
 * @property int|null $stock
 * @property string|null $image
 * @property array|null $gallery
 * @property bool $is_active
 * @property bool $is_featured
 * @property string|null $status
 * @property \Carbon\Carbon|null $published_at
 * @property string|null $meta_title
 * @property string|null $meta_description
 * @property array|null $badges
 * @property array|null $nutrition_facts
 * @property array|null $nutrition_info
 * @property string|null $label_copy
 * @property array|null $specifications
 * @property string|null $storage_instructions
 * @property string|null $safety_advice
 * @property string|null $allergen_note
 * @property array|null $customer_care
 * @property array|null $regulatory_info
 * @property string|null $barcode_url
 * @property string|null $qr_code_url
 * @property-read string|null $image_url
 * @property-read array $gallery_urls
 * @property \Carbon\Carbon $created_at
 * @property \Carbon\Carbon $updated_at
 */
class Product extends Model
{
    use HasFactory;

    protected $fillable = [
        'category_id',
        'product_category_id',
        'name',
        'slug',
        'sku',
        'description',
        'short_description',
        'content',
        'ingredients',
        'age_group',
        'type',
        'price',
        'sale_price',
        'stock',
        'image',
        'gallery',
        'is_active',
        'is_featured',
        'status',
        'published_at',
        'meta_title',
        'meta_description',
        'badges',
        'nutrition_facts',
        'nutrition_info',
        'label_copy',
        'specifications',
        'storage_instructions',
        'safety_advice',
        'allergen_note',
        'customer_care',
        'regulatory_info',
        'barcode_url',
        'qr_code_url',
    ];

    protected $casts = [
        'price'           => 'decimal:2',
        'sale_price'      => 'decimal:2',
        'gallery'         => 'array',
        'badges'          => 'array',
        'nutrition_facts' => 'array',
        'nutrition_info'  => 'array',
        'specifications'  => 'array',
        'customer_care'   => 'array',
        'regulatory_info' => 'array',
        'published_at'    => 'datetime',
        'is_active'       => 'boolean',
        'is_featured'     => 'boolean',
    ];

    protected $appends = ['image_url', 'gallery_urls'];

    public function getImageUrlAttribute(): ?string
    {
        if (empty($this->image)) {
            return null;
        }

        return $this->resolveAssetUrl($this->image);
    }

    public function getGalleryUrlsAttribute(): array
    {
        $gallery = $this->gallery ?? [];

        return array_map(function ($path) {
            return $this->resolveAssetUrl($path);
        }, $gallery);
    }

    public function getRouteKeyName(): string
    {
        return 'slug';
    }

    public function category()
    {
        return $this->belongsTo(Category::class);
    }

    public function productCategory()
    {
        return $this->belongsTo(ProductCategory::class);
    }

    public function orderItems()
    {
        return $this->hasMany(OrderItem::class);
    }

    public function pricingPlans()
    {
        return $this->belongsToMany(PricingPlan::class, 'pricing_plan_products')
            ->withPivot('quantity')
            ->withTimestamps();
    }

    /** Effective selling price (sale_price if set, otherwise price) */
    public function getEffectivePriceAttribute(): float
    {
        return (float) ($this->sale_price ?? $this->price ?? 0);
    }

    public function reviews()
    {
        return $this->hasMany(ProductReview::class);
    }

    public function approvedReviews()
    {
        return $this->hasMany(ProductReview::class)->where('moderation_status', 'approved');
    }

    public function wishlists()
    {
        return $this->hasMany(Wishlist::class);
    }

    public function cartItems()
    {
        return $this->hasMany(CartItem::class);
    }

    private function resolveAssetUrl(?string $path): ?string
    {
        if (empty($path)) {
            return null;
        }

        if (str_starts_with($path, 'http://') || str_starts_with($path, 'https://')) {
            return $path;
        }

        if (str_starts_with($path, '/assets/')) {
            return asset(ltrim($path, '/'));
        }

        if (str_starts_with($path, 'assets/')) {
            return asset($path);
        }

        if (str_starts_with($path, '/storage/')) {
            return asset(ltrim($path, '/'));
        }

        /** @var \Illuminate\Filesystem\FilesystemAdapter $disk */
        $disk = Storage::disk('public');
        return $disk->url($path);
    }
}

File: mobile-app/lib/src/models/product.dart
import '../config/app_config.dart';

class Product {
  final int id;
  final String name;
  final String slug;
  final String? sku;
  final String? description;
  final String? shortDescription;
  final String? content;
  final String? ingredients;
  final String? ageGroup;
  final String? type;
  final double price;
  final double? salePrice;
  final int stock;
  final String? image;
  final String? imageUrl;
  final List<String> gallery;
  final List<String> galleryUrls;
  final bool isActive;
  final bool isFeatured;
  final List<String> badges;
  final Map<String, dynamic>? nutritionFacts;
  final Map<String, dynamic>? nutritionInfo;
  final ProductCategory? productCategory;

  const Product({
    required this.id,
    required this.name,
    required this.slug,
    this.sku,
    this.description,
    this.shortDescription,
    this.content,
    this.ingredients,
    this.ageGroup,
    this.type,
    required this.price,
    this.salePrice,
    this.stock = 0,
    this.image,
    this.imageUrl,
    this.gallery = const [],
    this.galleryUrls = const [],
    this.isActive = true,
    this.isFeatured = false,
    this.badges = const [],
    this.nutritionFacts,
    this.nutritionInfo,
    this.productCategory,
  });

  double get effectivePrice => salePrice ?? price;
  bool get isOnSale => salePrice != null && salePrice! < price;
  bool get inStock => stock > 0;
  String? get displayImageUrl => _normalizeImageUrl(imageUrl ?? image);

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: _toInt(json['id']),
      name: json['name'] as String? ?? '',
      slug: json['slug'] as String? ?? '',
      sku: json['sku'] as String?,
      description: json['description'] as String?,
      shortDescription: json['short_description'] as String?,
      content: json['content'] as String?,
      ingredients: json['ingredients'] as String?,
      ageGroup: json['age_group'] as String?,
      type: json['type'] as String?,
      price: _toDouble(json['price']),
      salePrice:
          json['sale_price'] != null ? _toDouble(json['sale_price']) : null,
      stock: _toInt(json['stock']),
      image: json['image'] as String?,
      imageUrl: json['image_url'] as String?,
      gallery: (json['gallery'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      galleryUrls: (json['gallery_urls'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      isActive: _toBool(json['is_active'], true),
      isFeatured: _toBool(json['is_featured'], false),
      badges: (json['badges'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          const [],
      nutritionFacts: json['nutrition_facts'] as Map<String, dynamic>?,
      nutritionInfo: json['nutrition_info'] as Map<String, dynamic>?,
      productCategory: json['product_category'] != null
          ? ProductCategory.fromJson(
              json['product_category'] as Map<String, dynamic>)
          : null,
    );
  }

  static double _toDouble(dynamic v) {
    if (v is double) return v;
    if (v is int) return v.toDouble();
    if (v is String) return double.tryParse(v) ?? 0;
    return 0;
  }

  static int _toInt(dynamic v) {
    if (v is int) return v;
    if (v is double) return v.toInt();
    if (v is String) return int.tryParse(v) ?? 0;
    return 0;
  }

  static bool _toBool(dynamic v, bool fallback) {
    if (v is bool) return v;
    if (v is int) return v != 0;
    if (v is String) {
      final normalized = v.toLowerCase();
      if (normalized == 'true' || normalized == '1') return true;
      if (normalized == 'false' || normalized == '0') return false;
    }
    return fallback;
  }

  static String? _normalizeImageUrl(String? raw) {
    if (raw == null || raw.trim().isEmpty) return null;

    final value = raw.trim();
    if (value.startsWith('http://') || value.startsWith('https://')) {
      return value;
    }

    if (value.startsWith('/')) {
      return '${AppConfig.siteBaseUrl}$value';
    }

    if (value.startsWith('storage/') ||
        value.startsWith('assets/') ||
        value.startsWith('cms-media/')) {
      return '${AppConfig.siteBaseUrl}/$value';
    }

    return '${AppConfig.siteBaseUrl}/storage/$value';
  }
}

class ProductCategory {
  final int id;
  final String name;
  final String slug;

  const ProductCategory(
      {required this.id, required this.name, required this.slug});

  factory ProductCategory.fromJson(Map<String, dynamic> json) {
    return ProductCategory(
      id: json['id'] as int,
      name: json['name'] as String? ?? '',
      slug: json['slug'] as String? ?? '',
    );
  }
}
